### PR TITLE
tools/ceph-dencoder: refactor to address segfault when compiled w/ clang

### PR DESCRIPTION
**Pull request by tchaikov**
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by tchaikov** [comment]
https://pulpito.ceph.com/kchai-2021-08-04_05:19:17-rados-wip-kefu-testing-2021-08-04-1022-distro-basic-smithi/

**Comment by rzarzynski** [comment]
I confirm the fix works on my dev environment!

Before
---------
```
[rzarzynski@o06 build]$ ./../src/test/encoding/check-generated.sh
Environment Variables Not All Set, Detected Build System CMake
Setting Environment Variables
checking ceph-dencoder generated test instances...
numgen type
./../src/test/encoding/check-generated.sh: line 93: 1375002 Segmentation fault      (core dumped) ceph-dencoder list_types
```
After
-------
```
[rzarzynski@o06 build]$ ./../src/test/encoding/check-generated.sh
Environment Variables Not All Set, Detected Build System CMake
Setting Environment Variables
checking ceph-dencoder generated test instances...
numgen type
3 ACLGrant
2 ACLGranteeType
2 ACLOwner
...
2 snaplink_t
2 sr_t
3 sstring_wrapper
3 string_snap_t
1 timespan_wrapper
3 utime_t
4 uuid_d
2 watch_info_t
2 watch_item_t
passed 2358 tests.
```

**Review by rzarzynski — COMMENTED** [review]
Basically LGTM with one question about legacy and some tiny, little nits.

File: src/tools/ceph-dencoder/common_types.cc [hunk 1]
```diff
@@ -25,7 +25,7 @@ using namespace std;
 
 #include "denc_registry.h"
 
-DENC_API void register_dencoders()
+DENC_API void register_dencoders(DencoderRegistry& registry)
```
[the remark this answers is not in the document]
Comment (tchaikov) [reply src/tools/ceph-dencoder/common_types.cc]: i think, plugins can be considered part the application. in general, they are not soversioned. and we do package this way, for instance, in `ceph.spec.in`

https://github.com/ceph/ceph/blob/0dedabe15a9aababb86f9a9a1c0b9b0e94e0729b/ceph.spec.in#L1571

and 

https://github.com/ceph/ceph/blob/0dedabe15a9aababb86f9a9a1c0b9b0e94e0729b/ceph.spec.in#L1592

are packaged in the very same package.

File: src/tools/ceph-dencoder/ceph_dencoder.cc [hunk 2]
```diff
@@ -147,78 +147,79 @@ int main(int argc, const char **argv)
 
   if (args.empty()) {
     cerr << "-h for help" << std::endl;
-    exit(1);
+    return 1;
```
[the remark this answers is not in the document]
Comment (tchaikov) [reply src/tools/ceph-dencoder/ceph_dencoder.cc:112]: this is not part of this change. we can have a follow-up change to remove this magic number.

File: src/tools/ceph-dencoder/ceph_dencoder.cc [hunk 3]
```diff
@@ -77,18 +77,18 @@ class DencoderPlugin {
...
 #endif
   }
-  int register_dencoders() {
+  int register_dencoders(DencoderRegistry& registry) {
     assert(mod);
-    using register_dencoders_t = void (*)();
-    const auto register_dencoders =
+    using register_dencoders_t = void (*)(DencoderRegistry&);
+    const auto do_register =
       reinterpret_cast<register_dencoders_t>(dlsym(mod, REGISTER_DENCODERS_FUNCTION.data()));
```
[the remark this answers is not in the document]
Comment (tchaikov) [reply src/tools/ceph-dencoder/ceph_dencoder.cc]: as explained in another comment in this PR, i think we can assume that the plugins are of the same version of the executable.

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2021-08-04_05:19:17-rados-wip-kefu-testing-2021-08-04-1022-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-08-04_15:35:42-rados-wip-kefu-testing-2021-08-04-1022-distro-basic-smithi/

File: src/tools/ceph-dencoder/ceph_dencoder.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (rzarzynski) [reply src/tools/ceph-dencoder/ceph_dencoder.cc:112]: Sure, entirely fine with that.

File: src/tools/ceph-dencoder/common_types.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (rzarzynski) [reply src/tools/ceph-dencoder/common_types.cc]: Thanks for the clarification!

**Review by rzarzynski — APPROVED** [review]

**Comment by wjwithagen** [comment]
@tchaikov 
I'm have trouble loading the dynamic libs for cephfs, rgw and osd...
Like:
```
44058: failed to dlopen("/home/wjw/master/build/lib/denc-mod-rgw.so"): /home/wjw/master/build/lib/denc-mod-rgw.so: invalid file format
```
Have to figure out why this is...

File: src/tools/ceph-dencoder/denc_plugin.h [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+#include "denc_registry.h"
+
+namespace fs = std::filesystem;
+
+class DencoderPlugin {
+  using dencoders_t = std::vector<std::pair<std::string, Dencoder*>>;
+public:
+  DencoderPlugin(const fs::path& path) {
+    mod = dlopen(path.c_str(), RTLD_NOW);
```
Comment (wjwithagen) [inline src/tools/ceph-dencoder/denc_plugin.h:13]: @tchaikov
FreeBSD likes/needs `RTLD_NODELETE` with de dlopen flags.

Comment (wjwithagen) [reply src/tools/ceph-dencoder/denc_plugin.h:13]: PR in #42703

**Pull request merged**